src/analysis/normed_space/hahn_banach/separation.lean

```diff
@@ -0,0 +1,211 @@
+  refine ⟨f, s, hf₁, _⟩,
+  suffices : f '' t ⊆ Ioi s,
+  { exact λ b hb, this ⟨b, ‹_›, rfl⟩ },
+  rw ←interior_Ici,
+  refine interior_maximal _ (this _ ht₃),
+  rintro _ ⟨_, _, rfl⟩,
+  exact hf₂ _ ‹_›,
+end
+
+/-- A version of the Hahn-Banach theorem: given disjoint convex sets `s`, `t` where `s` is compact
+and `t` is closed, there is a continuous linear functional which strongly separates them. -/
+theorem geometric_hahn_banach_compact_closed {s t : set E} (hs₁ : convex ℝ s) (hs₂ : is_compact s)
+  (ht₁ : convex ℝ t) (ht₂ : is_closed t) (disj : disjoint s t) :
+  ∃ (f : E →L[ℝ] ℝ) (u v : ℝ), (∀ a ∈ s, f a < u) ∧ u < v ∧ ∀ b ∈ t, v < f b :=
+begin
+  obtain rfl | hs := s.eq_empty_or_nonempty,
+  { exact ⟨0, -2, -1, by simp, by norm_num, λ b hb, by norm_num⟩ },
+  tactic.unfreeze_local_instances,
+  obtain rfl | ht := t.eq_empty_or_nonempty,
+  { exact ⟨0, 1, 2, λ a ha, by norm_num, by norm_num, by simp⟩ },
+  obtain ⟨U, V, hU, hV, hU₁, hV₁, sU, tV, disj'⟩ :=
+    exists_disjoint_open_convexes hs₁ hs₂ ht₁ ht₂ disj,
+  obtain ⟨f, u, hf₁, hf₂⟩ := geometric_hahn_banach_open_open hU₁ hU hV₁ hV disj',
+  obtain ⟨x, hx₁, hx₂⟩ := hs₂.exists_forall_ge hs f.continuous.continuous_on,
+  have : Sup (f '' s) = f x,
+  { apply le_antisymm (cSup_le (hs.image f) (by simpa)),
+    exact le_cSup ⟨f x, by simpa [upper_bounds]⟩ ⟨_, hx₁, rfl⟩ },
+  have : f x < u := hf₁ x (sU hx₁),
+  exact ⟨f, (f x + u)/2, u, λ a ha, by linarith [hx₂ a ha], by linarith, λ b hb, hf₂ b (tV hb)⟩,
+end
+
+/-- A version of the Hahn-Banach theorem: given disjoint convex sets `s`, `t` where `s` is closed,
+and `t` is compact, there is a continuous linear functional which strongly separates them. -/
+theorem geometric_hahn_banach_closed_compact (hs₁ : convex ℝ s) (hs₂ : is_closed s)
+  (ht₁ : convex ℝ t) (ht₂ : is_compact t) (disj : disjoint s t) :
+  ∃ (f : E →L[ℝ] ℝ) (u v : ℝ), (∀ a ∈ s, f a < u) ∧ u < v ∧ ∀ b ∈ t, v < f b :=
+let ⟨f, s, t, hs, st, ht⟩ := geometric_hahn_banach_compact_closed ht₁ ht₂ hs₁ hs₂ disj.symm in
+⟨-f, -t, -s, by simpa using ht, by simpa using st, by simpa using hs⟩
+
+theorem geometric_hahn_banach_point_closed (ht₁ : convex ℝ t) (ht₂ : is_closed t) (disj : x ∉ t) :
+  ∃ (f : E →L[ℝ] ℝ) (u : ℝ), f x < u ∧ ∀ b ∈ t, u < f b :=
+let ⟨f, u, v, ha, hst, hb⟩ := geometric_hahn_banach_compact_closed (convex_singleton x)
+  is_compact_singleton ht₁ ht₂ (disjoint_singleton_left.2 disj)
+  in ⟨f, v, lt_trans (ha x (mem_singleton _)) hst, hb⟩
+
+theorem geometric_hahn_banach_closed_point {s : set E} {x : E} (hs₁ : convex ℝ s)
+  (hs₂ : is_closed s) (disj : x ∉ s) :
+  ∃ (f : E →L[ℝ] ℝ) (u : ℝ), (∀ a ∈ s, f a < u) ∧ u < f x :=
+let ⟨f, s, t, ha, hst, hb⟩ := geometric_hahn_banach_closed_compact hs₁ hs₂ (convex_singleton x)
+  is_compact_singleton (disjoint_singleton_right.2 disj)
+  in ⟨f, s, ha, lt_trans hst (hb x (mem_singleton _))⟩
+
+theorem geometric_hahn_banach_point_point {x y : E} (hxy : x ≠ y) : ∃ (f : E →L[ℝ] ℝ), f x < f y :=
```

Comment:
Could you please cross-reference this with the lemma [`normed_space.eq_iff_forall_dual_eq`](https://leanprover-community.github.io/mathlib_docs/analysis/normed_space/dual.html#normed_space.eq_iff_forall_dual_eq)?  (a reference in the docstring of each to the other)